--- busybox-1.23.2/networking/wget.c.orig	2015-03-23 04:06:55.000000000 +0100
+++ busybox-1.23.2/networking/wget.c	2015-03-23 16:17:08.521979589 +0100
@@ -15,6 +15,7 @@
 //usage:       "	[--header 'header: value'] [-Y|--proxy on/off] [-P DIR]\n"
 /* Since we ignore these opts, we don't show them in --help */
 /* //usage:    "	[--no-check-certificate] [--no-cache]" */
+//usage:       "	[--post-data DATA] [--post-file FILE]\n"
 //usage:       "	[-U|--user-agent AGENT]" IF_FEATURE_WGET_TIMEOUT(" [-T SEC]") " URL..."
 //usage:	)
 //usage:	IF_NOT_FEATURE_WGET_LONG_OPTIONS(
@@ -104,6 +105,7 @@
 	char *dir_prefix;
 #if ENABLE_FEATURE_WGET_LONG_OPTIONS
 	char *post_data;
+	char *post_file;
 	char *extra_headers;
 	unsigned char user_headers; /* Headers mentioned by the user */
 #endif
@@ -148,6 +150,7 @@
 	WGET_OPT_PASSIVE    = (1 << 9),
 	WGET_OPT_HEADER     = (1 << 10) * ENABLE_FEATURE_WGET_LONG_OPTIONS,
 	WGET_OPT_POST_DATA  = (1 << 11) * ENABLE_FEATURE_WGET_LONG_OPTIONS,
+	WGET_OPT_POST_FILE  = (1 << 12) * ENABLE_FEATURE_WGET_LONG_OPTIONS,
 };
 
 enum {
@@ -872,7 +875,7 @@
 				target.path);
 		} else {
 			SENDFMT(sfp, "%s /%s HTTP/1.1\r\n",
-				(option_mask32 & WGET_OPT_POST_DATA) ? "POST" : "GET",
+				(option_mask32 & WGET_OPT_POST_DATA || option_mask32 & WGET_OPT_POST_FILE) ? "POST" : "GET",
 				target.path);
 		}
 		if (!USR_HEADER_HOST)
@@ -913,6 +916,27 @@
 				"%s",
 				(int) strlen(G.post_data), G.post_data
 			);
+		} else if (option_mask32 & WGET_OPT_POST_FILE) {
+			FILE *ffp;
+			int c, file_size;
+
+			ffp = fopen(G.post_file, "r");
+			if (ffp == NULL) {
+				bb_error_msg_and_die("can not open file: %s", G.post_file);
+			} else {
+				fseek(ffp, 0L, SEEK_END);
+				file_size = ftell(ffp);
+				fprintf(sfp,
+					"Content-Type: application/x-www-form-urlencoded\r\n"
+					"Content-Length: %u\r\n"
+					"\r\n",
+					file_size
+				);
+				fseek(ffp, 0L, SEEK_SET);
+				while ((c = fgetc(ffp)) != EOF)
+					fputc(c, sfp);
+				fclose(ffp);
+			}
 		} else
 #endif
 		{
@@ -1110,10 +1134,11 @@
 		"passive-ftp\0"      No_argument       "\xff"
 		"header\0"           Required_argument "\xfe"
 		"post-data\0"        Required_argument "\xfd"
+		"post-file\0"        Required_argument "\xfc"
 		/* Ignored (we don't do ssl) */
-		"no-check-certificate\0" No_argument   "\xfc"
+		"no-check-certificate\0" No_argument   "\xfb"
 		/* Ignored (we don't support caching) */
-		"no-cache\0"         No_argument       "\xfb"
+		"no-cache\0"         No_argument       "\xfa"
 		;
 #endif
 
@@ -1143,6 +1168,7 @@
 		NULL /* -t RETRIES */
 		IF_FEATURE_WGET_LONG_OPTIONS(, &headers_llist)
 		IF_FEATURE_WGET_LONG_OPTIONS(, &G.post_data)
+		IF_FEATURE_WGET_LONG_OPTIONS(, &G.post_file)
 	);
 	argv += optind;
 
